File: java/agentbay/src/main/java/com/aliyun/agentbay/network/Network.java
package com.aliyun.agentbay.network;

import com.aliyun.agentbay.AgentBay;
public class Network {
    private AgentBay agentBay;

    public Network(AgentBay agentBay) {
        this.agentBay = agentBay;
    }

    public NetworkResult getNetworkBindToken() {
        return getNetworkBindToken(null);
    }

    public NetworkResult getNetworkBindToken(String networkId) {
        try {
            com.aliyun.wuyingai20250506.models.CreateNetworkRequest request =
                new com.aliyun.wuyingai20250506.models.CreateNetworkRequest();
            request.setAuthorization("Bearer " + agentBay.getApiKey());

            if (networkId != null && !networkId.isEmpty()) {
                request.setNetworkId(networkId);
            }

            com.aliyun.wuyingai20250506.models.CreateNetworkResponse response =
                agentBay.getClient().createNetwork(request);

            if (response == null || response.getBody() == null) {
                return new NetworkResult(
                    "",
                    false,
                    null,
                    null,
                    "Invalid response from CreateNetwork API"
                );
            }

            com.aliyun.wuyingai20250506.models.CreateNetworkResponseBody body = response.getBody();
            String requestId = body.getRequestId() != null ? body.getRequestId() : "";

            if (body.getSuccess() == null || !body.getSuccess()) {
                String errorMsg = body.getMessage() != null ? body.getMessage() : "Unknown error";
                if (body.getCode() != null) {
                    errorMsg = "[" + body.getCode() + "] " + errorMsg;
                }
                return new NetworkResult(
                    requestId,
                    false,
                    null,
                    null,
                    errorMsg
                );
            }

            if (body.getData() == null) {
                return new NetworkResult(
                    requestId,
                    false,
                    null,
                    null,
                    "Network data not found in response"
                );
            }

            String createdNetworkId = body.getData().getNetworkId();
            String networkToken = body.getData().getNetworkToken();
            return new NetworkResult(
                requestId,
                true,
                createdNetworkId,
                networkToken,
                ""
            );

        } catch (com.aliyun.tea.TeaException e) {
            String errorStr = e.getMessage();
            String requestId = "";
            if (e.getData() != null && e.getData().get("RequestId") != null) {
                requestId = e.getData().get("RequestId").toString();
            }
            return new NetworkResult(
                requestId,
                false,
                null,
                null,
                "Failed to create network: " + errorStr
            );
        } catch (Exception e) {
            return new NetworkResult(
                "",
                false,
                null,
                null,
                "Failed to create network: " + e.getMessage()
            );
        }
    }

    /**
     * Deprecated: use getNetworkBindToken().
     */
    public NetworkResult create() {
        return getNetworkBindToken();
    }

    /**
     * Deprecated: use getNetworkBindToken(String).
     */
    public NetworkResult create(String networkId) {
        return getNetworkBindToken(networkId);
    }

    public NetworkStatusResult describe(String networkId) {
        if (networkId == null || networkId.isEmpty()) {
            return new NetworkStatusResult(
                "",
                false,
                false,
                "network_id is required"
            );
        }

        try {
            com.aliyun.wuyingai20250506.models.DescribeNetworkRequest request =
                new com.aliyun.wuyingai20250506.models.DescribeNetworkRequest();
            request.setAuthorization("Bearer " + agentBay.getApiKey());
            request.setNetworkId(networkId);

            com.aliyun.wuyingai20250506.models.DescribeNetworkResponse response =
                agentBay.getClient().describeNetwork(request);

            if (response == null || response.getBody() == null) {
                return new NetworkStatusResult(
                    "",
                    false,
                    false,
                    "Invalid response from DescribeNetwork API"
                );
            }

            com.aliyun.wuyingai20250506.models.DescribeNetworkResponseBody body = response.getBody();
            String requestId = body.getRequestId() != null ? body.getRequestId() : "";

            if (body.getSuccess() == null || !body.getSuccess()) {
                String errorMsg = body.getMessage() != null ? body.getMessage() : "Unknown error";
                if (body.getCode() != null) {
                    errorMsg = "[" + body.getCode() + "] " + errorMsg;
                }
                return new NetworkStatusResult(
                    requestId,
                    false,
                    false,
                    errorMsg
                );
            }

            boolean online = false;
            if (body.getData() != null && body.getData().getOnline() != null) {
                online = body.getData().getOnline();
            }
            return new NetworkStatusResult(
                requestId,
                true,
                online,
                ""
            );

        } catch (com.aliyun.tea.TeaException e) {
            String errorStr = e.getMessage();
            String requestId = "";
            if (e.getData() != null && e.getData().get("RequestId") != null) {
                requestId = e.getData().get("RequestId").toString();
            }

            if (errorStr != null && errorStr.contains("NotFound")) {
                return new NetworkStatusResult(
                    requestId,
                    false,
                    false,
                    "Network " + networkId + " not found"
                );
            }
            return new NetworkStatusResult(
                requestId,
                false,
                false,
                "Failed to describe network: " + errorStr
            );
        } catch (Exception e) {
            return new NetworkStatusResult(
                "",
                false,
                false,
                "Failed to describe network: " + e.getMessage()
            );
        }
    }
}

File: golang/pkg/agentbay/network.go
package agentbay

import (
	"fmt"
	"strings"
	"time"

	"github.com/alibabacloud-go/tea/tea"
	mcp "github.com/aliyun/wuying-agentbay-sdk/golang/api/client"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
)

// NetworkResult wraps network create result and RequestID.
type NetworkResult struct {
	models.ApiResponse
	Success      bool
	NetworkId    string
	NetworkToken string
	ErrorMessage string
}

// NetworkStatusResult wraps network status result and RequestID.
type NetworkStatusResult struct {
	models.ApiResponse
	Success      bool
	Online       bool
	ErrorMessage string
}

// NetworkService provides methods to manage networks in the AgentBay cloud environment.
type NetworkService struct {
	AgentBay *AgentBay
}

// GetNetworkBindToken creates a network (or reuses provided networkId) and returns networkId + networkToken.
// If networkId is empty, a new network will be created.
func (ns *NetworkService) GetNetworkBindToken(networkId string) (*NetworkResult, error) {
	request := &mcp.CreateNetworkRequest{
		Authorization: tea.String("Bearer " + ns.AgentBay.APIKey),
	}
	if networkId != "" {
		request.NetworkId = tea.String(networkId)
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.CreateNetworkResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = ns.AgentBay.Client.CreateNetwork(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}
	requestID := models.ExtractRequestID(resp)
	if err != nil {
		return &NetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: fmt.Sprintf("Failed to create network: %v", err),
		}, nil
	}

	if resp == nil || resp.Body == nil {
		return &NetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: "Invalid response from CreateNetwork API",
		}, nil
	}

	body := resp.Body
	if body.Success == nil || !*body.Success {
		errorMsg := "Unknown error"
		if body.Message != nil {
			errorMsg = *body.Message
		}
		if body.Code != nil {
			errorMsg = fmt.Sprintf("[%s] %s", *body.Code, errorMsg)
		}
		return &NetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: errorMsg,
		}, nil
	}

	if body.Data == nil {
		return &NetworkResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			ErrorMessage: "Network data not found in response",
		}, nil
	}

	createdNetworkID := ""
	if body.Data.NetworkId != nil {
		createdNetworkID = *body.Data.NetworkId
	}
	networkToken := ""
	if body.Data.NetworkToken != nil {
		networkToken = *body.Data.NetworkToken
	}

	return &NetworkResult{
		ApiResponse:  models.ApiResponse{RequestID: requestID},
		Success:      true,
		NetworkId:    createdNetworkID,
		NetworkToken: networkToken,
		ErrorMessage: "",
	}, nil
}

// Create is kept for backward compatibility. Prefer GetNetworkBindToken.
func (ns *NetworkService) Create(networkId string) (*NetworkResult, error) {
	return ns.GetNetworkBindToken(networkId)
}

// Describe queries network status (online/offline).
func (ns *NetworkService) Describe(networkId string) (*NetworkStatusResult, error) {
	if networkId == "" {
		return &NetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: ""},
			Success:      false,
			Online:       false,
			ErrorMessage: "network_id is required",
		}, nil
	}

	request := &mcp.DescribeNetworkRequest{
		Authorization: tea.String("Bearer " + ns.AgentBay.APIKey),
		NetworkId:     tea.String(networkId),
	}

	maxAttempts := 3
	delay := 200 * time.Millisecond
	var resp *mcp.DescribeNetworkResponse
	var err error
	for attempt := 1; attempt <= maxAttempts; attempt++ {
		resp, err = ns.AgentBay.Client.DescribeNetwork(request)
		if err == nil {
			break
		}
		if attempt < maxAttempts {
			errStr := err.Error()
			if strings.Contains(errStr, "ServiceUnavailable") || strings.Contains(errStr, "503") {
				time.Sleep(delay)
				delay *= 2
				continue
			}
		}
	}
	requestID := models.ExtractRequestID(resp)
	if err != nil {
		errorStr := err.Error()
		if strings.Contains(errorStr, "NotFound") {
			return &NetworkStatusResult{
				ApiResponse:  models.ApiResponse{RequestID: requestID},
				Success:      false,
				Online:       false,
				ErrorMessage: fmt.Sprintf("Network %s not found", networkId),
			}, nil
		}
		return &NetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: fmt.Sprintf("Failed to describe network: %v", err),
		}, nil
	}

	if resp == nil || resp.Body == nil {
		return &NetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: "Invalid response from DescribeNetwork API",
		}, nil
	}

	body := resp.Body
	if body.Success == nil || !*body.Success {
		errorMsg := "Unknown error"
		if body.Message != nil {
			errorMsg = *body.Message
		}
		if body.Code != nil {
			errorMsg = fmt.Sprintf("[%s] %s", *body.Code, errorMsg)
		}
		return &NetworkStatusResult{
			ApiResponse:  models.ApiResponse{RequestID: requestID},
			Success:      false,
			Online:       false,
			ErrorMessage: errorMsg,
		}, nil
	}

	online := false
	if body.Data != nil && body.Data.Online != nil {
		online = *body.Data.Online
	}
	return &NetworkStatusResult{
		ApiResponse:  models.ApiResponse{RequestID: requestID},
		Success:      true,
		Online:       online,
		ErrorMessage: "",
	}, nil
}
